cluster-autoscaler/core/scale_down.go

```diff
@@ -554,9 +559,23 @@ func (sd *ScaleDown) chooseCandidates(nodes []*apiv1.Node) ([]*apiv1.Node, []*ap
 	return currentCandidates, currentNonCandidates
 }
 
-// TryToScaleDown tries to scale down the cluster. It returns ScaleDownResult indicating if any node was
+func (sd *ScaleDown) mapNodesToStatusScaleDownNodes(nodes []*apiv1.Node, nodeGroups map[string]cloudprovider.NodeGroup, evictedPodLists map[string][]*apiv1.Pod) []*status.ScaleDownNode {
+	var result []*status.ScaleDownNode
+	for _, node := range nodes {
+		result = append(result, &status.ScaleDownNode{
+			Node:        node,
+			NodeGroup:   nodeGroups[node.Name],
+			UtilInfo:    sd.nodeUtilizationMap[node.Name],
+			EvictedPods: evictedPodLists[node.Name],
+		})
+	}
+	return result
+}
+
+// TryToScaleDown tries to scale down the cluster. It returns a result inside a ScaleDownStatus indicating if any node was
 // removed and error if such occurred.
-func (sd *ScaleDown) TryToScaleDown(allNodes []*apiv1.Node, pods []*apiv1.Pod, pdbs []*policyv1.PodDisruptionBudget, currentTime time.Time) (ScaleDownResult, errors.AutoscalerError) {
+func (sd *ScaleDown) TryToScaleDown(allNodes []*apiv1.Node, pods []*apiv1.Pod, pdbs []*policyv1.PodDisruptionBudget, currentTime time.Time) (*status.ScaleDownStatus, errors.AutoscalerError) {
+	scaleDownStatus := &status.ScaleDownStatus{NodeDeleteResults: sd.nodeDeleteStatus.DrainNodeDeleteResults()}
```

Comment:
I sort of agree with the idea of draining node delete results only at the beginning of scale-down (i.e., when we know we'll be processing a scale down status), but I was wondering what will happen if:
- we start draining a node,
- a schedulable pod appears (i.e., schedulable in simulation, but not schedulable according to scheduler - it happens from time to time), or there's just a large churn of pods for a long while,
- node drain completes, but we don't enter TryToScaleDown because of the above condition for ~hours, so we don't process the results for this time. I assume it's up to ScaleDownStatusProcessor to know the deletion results it was expecting are stale, but it seems a pity it can't access them even though they've been ready this whole time.

Or:
- we start draining a node,
- it completes,
- next run of TryToScaleDown returns an error, and we exit from RunOnce without running ScaleDownStatus through ScaleDownStatusProcessor. However, the deletion results are lost without ever being processed.

Reply:
I've changed the processor invocation in RunOnce() to deal with the second scenario, but the first one looks out of scope of this PR.